Add migration guide from the Apollo CLI -> Rover (#425)
apollographql/apollo-cli: docs/source/migration.md

---
title: 'Migrating from the Apollo CLI'
sidebar_title: 'Migrating from the Apollo CLI'
---

The [Apollo CLI](https://github.com/apollographql/apollo-tooling) is Apollo's previous CLI tool for managing graphs with [Apollo Studio's graph registry](https://www.apollographql.com/docs/intro/platform/#3-manage-your-graph-with-apollo-studio).

This guide will help you migrate from the Apollo CLI to Rover, providing a guide for many of the key differences in Rover and the Apollo CLI, along with examples of the most common workflows from the Apollo CLI.

> If you haven't already read the [Conventions](./essentials) doc, start there, since this doc builds off it and assumes you have already read it. It will also be helpful to [install](./getting-started) and [configure](./configuring) Rover before reading through this guide.

Reviewer: > [Conventions](./essentials)

^ should we rename the `essentials` to `conventions` or rename the link here?

> **Rover is in active, rapid development.** Rover commands are subject to breaking changes without notice. Until the release of version 0.1.0 or higher, we do not yet recommend integrating Rover in critical production systems.
>
> For details, see [Public preview](./#public-preview).

## Configuring

> Rover's approach to [configuration](./configuring) is much more granular and less abstracted than the Apollo CLI. Most configuration options in Rover are configured as flags on a command, rather than items in a config file. The only config files that Rover uses are hidden config files created by the `config auth` command and another as input to [`supergraph compose`](./supergraphs#configuration).

Reviewer: Perhaps it's worth noting what the old config format looks like/the info it contains and then map each of those values to how you'd configure it in rover?

Author: I'll see what I can do for that. I had the same thought, but there's a bunch of edge cases with the old config. I think I can figure something out though 😄

## Limitations of Rover

Not all of the Apollo CLI's features have a compatible feature in Rover, but many of the most popular features are supported in Rover.

For the sake of maintainability, roadmap considerations, and time, Rover has intentionally left off some familiar features of the Apollo CLI.

### Client commands

Rover's focus is around providing an excellent graph management experience. For that reason, we have intentionally left out commands that may be useful for developers working on client projects like `client:codegen` and `client:check`.

### Globs

The Apollo CLI used globs extensively to support using multiple local files and even automatic discovery of files in a directory tree. While this was helpful in a lot of cases, the globbing strategy in the Apollo CLI resulted in a lot of issues and confusion. Rover has intentionally left globs off of file specifiers for initial release.

As a workaround, you may be able to use `cat` to combine multiple files, and pass them to Rover with [stdin](./essentials#io)

### Machine-readable output

> In the Apollo CLI, many commands supported alternate output formatting options like `--json` and `--markdown`. We do intend on adding structured output that can be machine readable, but Rover currently does not support it.

Reviewer: > We do intend on adding structured output that can be machine readable, but Rover currently does not support it.

^ i'm a bit wary of mentioning future plans in a doc like this as it can lead to disappointment if it takes too long or never really happens. i think we'll definitely do `--json` but not as sure about `--markdown`

```suggestion
In the Apollo CLI, many commands supported alternate output formatting options like `--json` and `--markdown`. This is not currently supported by Rover.
```

## Examples

> Note: these examples assume an environment variable or config file has been set up to authenticate the Apollo CLI, and the `rover config auth` command has been run to authenticate Rover.

### Fetching a graph's schema from Apollo's graph registry

```bash
# with Apollo
# automatically outputs to schema.graphql
apollo client:download-schema --graph my-graph --variant prod

# with Rover
# automatically outputs to stdout. Can redirect to schema.graphql
rover graph fetch my-graph@prod > schema.graphql
```

### Fetching a graph's schema from introspection

```bash
# with Apollo
# automatically outputs to schema.graphql
# can ONLY output introspection results in JSON
apollo service:download --endpoint http://localhost:4000

# with Rover
# automatically outputs to stdout. Can redirect to schema.graphql
# can ONLY output SDL
rover graph introspect http://localhost:4000
```

### Publishing a monolithic schema to the Apollo graph registry

```bash
# with Apollo
apollo service:push --graph my-graph --variant prod --endpoint http://localhost:4000

# with Rover

> rover graph introspect https://localhost:4000 | rover graph publish my-graph@prod --schema -

Reviewer: i love love love that we can do this so much
